dev/NavigationView/NavigationView.cpp

```diff
@@ -1594,6 +1734,222 @@ void NavigationView::UpdateVisualStateForDisplayModeGroup(const winrt::Navigatio
+        if (auto nvi = sender.try_as<winrt::NavigationViewItem>())
+        {
+            HandleKeyEventForNavigationViewItem(nvi, args);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemKeyDown(const winrt::IInspectable& sender, const winrt::KeyRoutedEventArgs& args)
+{
+    if (auto nvi = sender.try_as<winrt::NavigationViewItem>())
+    {
+        HandleKeyEventForNavigationViewItem(nvi, args);
+    }
+}
+
+void NavigationView::HandleKeyEventForNavigationViewItem(const winrt::NavigationViewItem& nvi, const winrt::KeyRoutedEventArgs& args)
+{
+    auto key = args.Key();
+    if (IsSettingsItem(nvi))
+    {
+        // Because ListViewItem eats the events, we only get these keys on KeyDown.
+        if (key == winrt::VirtualKey::Space ||
+            key == winrt::VirtualKey::Enter)
+        {
+            args.Handled(true);
+            OnSettingsInvoked();
+        }
+    }
+    else
+    {
+        switch (key)
+        {
+        case winrt::VirtualKey::Enter:
+        case winrt::VirtualKey::Space:
+            args.Handled(true);
+            OnNavigationViewItemInvoked(nvi);
+            break;
+        case winrt::VirtualKey::Home:
+            args.Handled(true);
+            KeyboardFocusFirstItemFromItem(nvi);
+            break;
+        case winrt::VirtualKey::End:
+            args.Handled(true);
+            KeyboardFocusLastItemFromItem(nvi);
+            break;
+        }
+    }
+}
+
+void NavigationView::KeyboardFocusFirstItemFromItem(const winrt::NavigationViewItemBase& nvib)
+{
+    auto const firstElement = [this, nvib]()
+    {
+        if (IsTopNavigationView())
+        {
+            if (IsContainerInOverflow(nvib))
+            {
+                return m_topNavRepeaterOverflowView.get().TryGetElement(0);
+            }
+            else
+            {
+                return m_topNavRepeater.get().TryGetElement(0);
+            }
+        }
+        return m_leftNavRepeater.get().TryGetElement(0);
+    }();
+
+    if (auto controlFirst = firstElement.try_as<winrt::Control>())
+    {
+        controlFirst.Focus(winrt::FocusState::Keyboard);
+    }
+}
+
+void NavigationView::KeyboardFocusLastItemFromItem(const winrt::NavigationViewItemBase& nvib)
+{
+    auto const ir = [this, nvib]()
+    {
+        if (IsTopNavigationView())
+        {
+            if (IsContainerInOverflow(nvib))
+            {
+                return m_topNavRepeaterOverflowView.get();
+            }
+            else
+            {
+                return m_topNavRepeater.get();
+            }
+        }
+        else
+        {
+            return m_leftNavRepeater.get();
+        }
+    }();
+
+    if (auto itemsSourceView = ir.ItemsSourceView())
+    {
+        auto lastIndex = itemsSourceView.Count() - 1;
+        if (auto lastElement = ir.TryGetElement(lastIndex))
+        {
+            if (auto controlLast = lastElement.try_as<winrt::Control>())
+            {
+                controlLast.Focus(winrt::FocusState::Keyboard);
```

Comment:
>Focus [](start = 28, length = 5)

Usually when calling Focus from a control, you'll want to use FocusState::Programmatic, which uses the last used input.